compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1955,6 +1964,410 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            // This shouldn't happen, but since this is only a diagnostic improvement, avoid breaking things.
+            return Err(expr);
+        }
+        let relevant_broken_predicate: ty::PredicateKind<'tcx> =
+            impl_predicates.predicates[impl_predicate_index].0.kind().skip_binder();
+
+        // We want to find which of the generics in the `impl_generics` are relevant to
+        // the broken obligation predicate.
+        // A generic is relevant if it is mentioned in the "original" predicate. If we can't narrow it down, treat them all as relevant.
+        let relevant_impl_generics = impl_generics.params.iter();
+
+        let relevant_impl_generics: Vec<&ty::GenericParamDef> = match relevant_broken_predicate {
+            ty::PredicateKind::Clause(ty::Clause::Trait(broken_trait)) => relevant_impl_generics
+                .filter(|&generic| {
+                    // Only retain generics that are mentioned in the (Self type) of this predicate:
+                    find_param_def_in_ty_walker(broken_trait.trait_ref.self_ty().walk(), generic)
+                })
+                .collect(),
+            _ => {
+                // Because we are missing information about which might be relevant, we assume that they all are.
+                relevant_impl_generics.collect()
+            }
+        };
+
+        enum PointableType<'tcx> {
+            Adt(ty::AdtDef<'tcx>),
+            Tuple,
+        }
+
+        // Note that there is no point in supporting "primitive" types like char/i32,
+        // since we cannot refine such a span any more anyway.
+        // Right now, only ADTs (struct/enum (variant) constructors) and tuples are supported by this refinement.
+        let (impl_self_ty, impl_self_ty_args) = match impl_self_ty.kind() {
+            ty::Adt(impl_self_ty_path, impl_self_ty_args) => (
+                PointableType::Adt(*impl_self_ty_path),
+                impl_self_ty_args.iter().collect::<Vec<_>>(),
+            ),
+            ty::Tuple(impl_self_ty_args) => (
+                PointableType::Tuple,
+                impl_self_ty_args.iter().map(|t| ty::GenericArg::from(t)).collect::<Vec<_>>(),
+            ),
+            _ => {
+                return Err(expr);
+            }
+        };
+
+        // We need to find which of the type's arguments are relevant to this obligation.
+        // For example, if we had an impl for `SomeTime<(A, C), B, Result<B, C>>` and the broken obligation
+        // was `B: Display` then we'd care about indexes `vec![1, 2]`, but if it was `C: PartialEq` then we'd
+        // care about index `vec![0, 2]`.
+        let relevant_ty_args_indices: Vec<usize> = impl_self_ty_args
+            .into_iter()
+            .enumerate()
+            .filter(|(_index, ty_arg)| {
+                relevant_impl_generics
+                    .iter()
+                    .any(|param| find_param_def_in_ty_walker(ty_arg.walk(), param))
```

Comment:
If you can't actually apply the comment I left below, then please at least pass in `ty_arg` rather than `ty_arg.walk()` here, if possible.